packages/sanity/src/desk/panes/document/DocumentPaneProvider.tsx

```diff
@@ -119,22 +114,38 @@ export const DocumentPaneProvider = memo((props: DocumentPaneProviderProps) => {
     params.path ? pathFromString(params.path) : []
   )
   const activeViewId = params.view || (views[0] && views[0].id) || null
-  const timeline = useMemo(
-    () => historyStore.getTimeline({publishedId: documentId, enableTrace: isDev}),
-    [documentId, historyStore]
-  )
   const [timelineMode, setTimelineMode] = useState<'since' | 'rev' | 'closed'>('closed')
-  // NOTE: this emits sync so can never be null
-  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-  const {historyController} = useMemoObservable(
-    () => historyStore.getTimelineController({client, documentId, documentType, timeline}),
-    [client, documentId, documentType, timeline]
-  )!
-
-  // @todo: this will now happen on each render, but should be refactored so it happens only when
-  // the `rev`, `since` or `historyController` values change.
-  historyController.setRange(params.since || null, params.rev || null)
-  const changesOpen = historyController.changesPanelActive()
+  const changesOpen = !!params.since
+
+  // Create a TimelineController and receive a subscribable BehaviorSubject, made accessible
+  // to all child components via DocumentPaneContext.
+  const {timelineController$} = useTimelineController({
+    documentId,
+    documentType,
+    rev: params.rev,
+    since: params.since,
+  })
+
+  // Subscribe to TimelineController changes and store internal state.
+  const [onOlderRevision, setOnOlderRevision] = useState(false)
+  const [revTime, setRevTime] = useState<Chunk | null>(null)
+  const [sinceAttributes, setSinceAttributes] = useState<Record<string, unknown> | null>(null)
+  const [timelineDisplayed, setTimelineDisplayed] = useState<Record<string, unknown> | null>(null)
+  const [timelineReady, setTimelineReady] = useState(false)
+  const [timelineError, setTimelineError] = useState<Error | null>(null)
+  useEffect(() => {
+    const subscription = timelineController$.subscribe({
+      error: (err) => setTimelineError(err),
+      next: (controller) => {
+        setOnOlderRevision(controller.onOlderRevision())
+        setRevTime(controller.revTime)
+        setSinceAttributes(controller.sinceAttributes())
+        setTimelineDisplayed(controller.displayed())
+        setTimelineReady(!['invalid', 'loading'].includes(controller.selectionState))
+      },
+    })
+    return () => subscription.unsubscribe()
+  }, [timelineController$])
```

Comment:
Plenty of room for improvement here, it's quite verbose but explicit. I tried implementing selectors here with `useSyncExternalStore`, though came up short.

Reply:
Agreed - but no rush as this can always be improved on later. LMK if you want to team up to look into how to improve it.